modules/docs/markdown/06-guides/smithy-transformations.md

```diff
@@ -0,0 +1,124 @@
+---
+sidebar_label: Smithy Transformations
+title: Smithy Transformations
+---
+
+There are times that you may want to transform the Smithy model being used by Smithy4s prior to code generation. In this guide, we will walk through exactly how you can accomplish this. As an example, we will show how you can remove members marked with a certain trait from structures inside of your Smithy model. However, you can use the same principles in this guide to accomplish whatever other transformations you may need.
+
+## Starting Smithy Model
+
+```smithy
+namespace preprocessed
+
+@trait(selector: "structure > member")
+structure removeBeforeCodegen {}
+
+structure MyStruct {
+    @required
+    name: String
+
+    @removeBeforeCodegen
+    id: String
+}
+```
+
+Here we have defined a trait `removeBeforeCodegen`. We have marked the `id` member of `MyStruct` with this trait. As such, we will implement a transformer which will lead to the model looking as follows:
+
+```smithy
+namespace preprocessed
+
+structure MyStruct {
+    @required
+    name: String
+}
+```
+
+This is the model that will ultimately be fed into the Smithy4s code generation tooling.
+
+## Implement the Transformer
+
+Now we will walk through the steps of actually implementing the Smithy model transformer. The `smithy-build` Java library provides an interface for these called `ProjectionTransformer`. Here is an example of a transformer that will remove the members marked with the `removeBeforeCodegen` trait as discussed above.
+
+```scala
+final class RemoveBeforeCodegenTransformation extends ProjectionTransformer {
+  def getName() = {
+    "RemoveBeforeCodegenTransformation"
+  }
+
+  def transform(x: TransformContext) = {
+    val toRemove = x
+      .getModel()
+      .getShapesWithTrait(new ToShapeId() {
+        override def toShapeId(): ShapeId =
+          ShapeId.from("preprocessed#removeBeforeCodegen")
+      })
+    x.getTransformer().removeShapes(x.getModel(), toRemove)
+  }
+}
+```
+
+Here we create a `ProjectionTransformer` named `RemoveBeforeCodegenTransformation`. Inside the `transform` method we remove all shapes that are marked with the `removeBeforeCodegen` trait and return the resultant model.
```

Comment:
nitpick: I'm not sure this line is that helpful as it says only _what_ we are doing, some of these should be pretty clear from the code itself.

Reply:
also if we're gonna keep it, I'd mention that the "name" of the transformer doesn't have to match the class name and is defined by the method we implement